devito/passes/clusters/aliases.py

```diff
@@ -154,17 +122,123 @@ def callbacks_sops(context, n):
     return processed
 
 
-def extract(cluster, rule1, model, template):
-    make = lambda: Scalar(name=template(), dtype=cluster.dtype).indexify()
+class Callbacks(object):
+
+    """
+    Interface for the callbacks needed by the CIRE loop. Each CIRE mode needs
+    to provide an implementation of these callbackes in a suitable subclass.
+    """
+
+    mode = None
+
+    def __new__(cls, context, n, min_cost):
+        min_cost = min_cost[cls.mode]
+        if callable(min_cost):
+            min_cost = min_cost(n)
+
+        return (partial(cls.extract, n, context, min_cost),
```

Comment:
Wait so this returns a tuple of class method wrapped in a partial? That feels a bit odd, can't it just be properties?
Also, I don't like partial (I know it's in FD and needs maybe rework), lambda are more efficient

Reply:
with this pattern you will never be able to instantiate the class, but you can still use inheritance. What you get instead are three callbacks. I think this is fine, if you had properties you would have to call them separately, while instead I want them as a unique block because **all* of them are necessary at the same time. The Callbacks class defines what in Java would have been an Interface -- something not instantiable but that still defines what the subclasses must implement

about `partial` -- I think you're too stressed about it, this is called once or twice per compilation and we're probably just talking about the 0.0001% of the turnaround time